pkg/reconciler/autoscaling/kpa/kpa.go

```diff
@@ -209,6 +217,55 @@ func (c *Reconciler) reconcileDecider(ctx context.Context, pa *pav1alpha1.PodAut
 	return decider, nil
 }
 
+func (c *Reconciler) updateContainerHealth(ctx context.Context, pa *pav1alpha1.PodAutoscaler) error {
+	logger := logging.FromContext(ctx)
+
+	if pa.Status.IsActive() {
+		pa.Status.MarkContainersHealthy()
+		return nil
+	}
+	ps, err := resourceutil.GetScaleResource(pa.Namespace, pa.Spec.ScaleTargetRef, c.PSInformerFactory)
+	if err != nil {
+		logger.Errorf("Error getting deployment: %v", err)
+		return err
+	}
+	// If a container keeps crashing (no active pods in the deployment although we want some)
+	if *ps.Spec.Replicas > 0 && ps.Status.Replicas == 0 {
+		pods, err := c.KubeClientSet.CoreV1().Pods(pa.Namespace).List(metav1.ListOptions{LabelSelector: metav1.FormatLabelSelector(ps.Spec.Selector)})
+		if err != nil {
+			logger.Errorf("Error getting pods: %v", err)
+			return err
+		} else if len(pods.Items) > 0 {
+			// Arbitrarily grab the very first pod, as they all should be crashing if one does.
+			pod := pods.Items[0]
+
+			// If pod cannot be scheduled then we expect the container status to be empty.
+			for _, cond := range pod.Status.Conditions {
+				if cond.Type == corev1.PodScheduled && cond.Status == corev1.ConditionFalse {
+					pa.Status.MarkPodUnscheduled(cond.Reason, cond.Message)
+					break
+				}
+			}
+
+			for _, status := range pod.Status.ContainerStatuses {
+				// Only check user container.
+				if len(ps.Spec.Template.Spec.Containers) > 0 && status.Name == ps.Spec.Template.Spec.Containers[0].Name {
+					if t := status.LastTerminationState.Terminated; t != nil {
+						logger.Infof("%s marking %s exiting with: %d/%s", pa.Name, status.Name, t.ExitCode, t.Message)
+						pa.Status.MarkContainerExiting(t.ExitCode, t.Message)
+					} else if w := status.State.Waiting; w != nil {
```

Comment:
Sorry for stepping on your toes with this :smile: https://github.com/knative/serving/pull/4195

I ended up adding the `hasDeploymentTimedOut` check because because being in a "Waiting" state isn't actually fatal. It's only a problem if the container is waiting indefinitely.

There was some code to check if a deployment was failing to progress so we could surface an error, but in the case of image pull errors, that check would always fail because of `IsActivationRequired`:
https://github.com/knative/serving/blob/ffeb54ad9c5029e3a04e49961900c8f4a55f06e1/pkg/reconciler/revision/reconcile_resources.go#L102

I'm wondering if we could update that code based on your changes to the pod autoscaler status to actually make sense. As it is, `MarkProgressDeadlineExceeded` will never be called if the image can't be pulled :/

Reply:
Thank you for pointing it out.  PA doesn't depend on `Deployment`, and I'm trying to translate the conditions to `PodScalable`, so that not to add a `Deployment` in PA.  Is there an equivalent check not using deployments?

Reply:
I'm using Reason to detect the problem.